### PR TITLE
Create a No-op stream in ClientCallImpl.start() if the Context has be…

### LINKED ISSUE #1253: ClientCallImpl operations race with Context cancellation.
With be6008666e9012a35f77defaab8e95730b81b838, we don't create the stream when the Context is cancelled, so the following request(), sendMessage(), halfClose() will encounter an IllegalStateException like:
```
java.lang.IllegalStateException: Not started
	at com.google.common.base.Preconditions.checkState(Preconditions.java:178)
	at io.grpc.internal.ClientCallImpl.request(ClientCallImpl.java:257)
	at io.grpc.ForwardingClientCall.request(ForwardingClientCall.java:50)
	at io.grpc.stub.ClientCalls.startCall(ClientCalls.java:199)
	at io.grpc.stub.ClientCalls.asyncUnaryRequestCall(ClientCalls.java:173)
	at io.grpc.stub.ClientCalls.futureUnaryCall(ClientCalls.java:135)
```

@louiscryan, FYI, I'll send you a PR to fix it soon.

### PR DESCRIPTION
**Pull request by madongfly**
…en cancelled.

Fixes #1253

@ejona86 @louiscryan Please take a look, thanks!

### CODE REVIEW

**Comment by ejona86** [comment]
@madongfly LGTM, but we should eventually use the same implementation with `DelayedStream.NOOP_CLIENT_STREAM` as well.

**Comment by madongfly** [comment]
@ejona86 I got the implementation in DelayedStream removed, PTAL, thanks!

File: core/src/main/java/io/grpc/internal/NoopClientStream.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import io.grpc.DecompressorRegistry;
+import io.grpc.Status;
+
+import java.io.InputStream;
+
+/**
+ * An implementation of {@link ClientStream} that silently does nothing for the operations.
+ */
+public class NoopClientStream implements ClientStream {
+  static NoopClientStream INSTANCE = new NoopClientStream();
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/NoopClientStream.java]: public?

**Comment by ejona86** [comment]
@madongfly, LGTM

File: core/src/main/java/io/grpc/internal/NoopClientStream.java (hunk 1 above)
Comment (madongfly) [inline core/src/main/java/io/grpc/internal/NoopClientStream.java]: Done.

**Pull request merged**